woocommerce/woocommerce-gateway-stripe
# Add off_session for recurring payments

This PR handles recurring subscription payments using PaymentIntents instead of Charges, and adds the `off_session` parameter to the payment intents API call.

This change affects subscription renewals that are either (1) renewed manually through wp-admin by the store owner as described in the testing steps, or (2) automatically on schedule. It does not affect subscription renewals manually triggered by the customer.

This change does not help with payment intents that still require authentication, even though `off_session` is being used.

Closes #938 and #931

## To test:

0. Install WC Subscriptions
1. Check out with a subscription
2. As a store owner / admin, go to the subscription edit page
3. Under "Order Actions" select "Process renewal" and click "update"
4. See that the subscription renewal goes through using a payment intent API call, as seen in the Stripe dashboard

Added: Try testing with the test card `4000002500003155`. A recurring payment will go through successfully on this branch, but not on `master`.

Can run the automated tests using `phpunit`.

## Things not asserted in automated tests:
- `wc_gateway_stripe_process_payment` action done (too hard to mock)
- `process_response()` is called
- not asserting the value of 'statement_descriptor', because it's hard to mock, so skipping for now

## Tests not written

- test_renewal_handles_charge_that_requires_authentication - this will come in another PR, solving #957
- test_renewal_handles_unsuccessful_charge_and_successful_retry - will come in a subsequent PR
- test_renewal_handles_unsuccessful_charge - will come in a subsequent PR
- test_renewal_throws_error_on_too_low_amount - may or not end up being tested for now
- test_renewal_throws_error_on_no_customer - may or not end up being tested for now

## Questions
- do we need to save payment intent to order? for now, I an doing so based on other payment intents
- @LCmry - why does WC Subscriptions send the `$amount` in the hook, instead of getting payment gateways to use the order amount?
